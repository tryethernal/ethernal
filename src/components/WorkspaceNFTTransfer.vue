<template>
    <v-container fluid>
        <h2 class="text-h6 font-weight-medium">NFT Transfers</h2>
        <v-divider class="my-4"></v-divider>
        <v-card>
            <v-card-text>
                <TokenTransfers
                    :transfers="transfers"
                    :withCount="false"
                    noDataText="No NFT transfers found."
                    :headers="headers"
                    :loading="loading"
                    :count="count"
                    @update:options="fetchTransfers"
                />
            </v-card-text>
        </v-card>
    </v-container>
</template>

<script setup>
import { ref, inject, shallowRef } from 'vue';
import TokenTransfers from './TokenTransfers.vue';

// Inject server instance
const $server = inject('$server');

// Reactive state
const transfers = ref([]);
const loading = ref(false);
const count = ref(0);

// Define table headers
const headers = shallowRef([
    { title: 'Tx Hash', key: 'transactionHash', sortable: false },
    { title: 'Method', key: 'methodDetails', sortable: false },
    { title: 'Age', key: 'timestamp'},
    { title: 'Block', key: 'blockNumber' },
    { title: 'From', key: 'src', sortable: false },
    { title: 'To', key: 'dst', sortable: false },
    { title: 'Type', key: 'type', sortable: false },
    { title: 'Token', key: 'token', sortable: false }
]);

// Default options for the data table
const defaultOptions = {
    page: 1,
    itemsPerPage: 10,
    sortBy: [{ key: 'timestamp', order: 'desc' }]
};

// Fetch transfers method
const fetchTransfers = (options = defaultOptions) => {
    loading.value = true;

    const params = {
        page: options.page,
        limit: options.itemsPerPage,
        orderBy: options.sortBy[0]?.key || defaultOptions.sortBy[0].key,
        order: options.sortBy[0]?.order || defaultOptions.sortBy[0].order,
        tokenTypes: ['erc721', 'erc1155']
    };

    $server.getWorkspaceTokenTransfers(params)
        .then(({ data }) => {
            transfers.value = data.items;
            count.value = data.items.length == options.itemsPerPage ?
                (options.page * options.itemsPerPage) + 1 :
                options.page * options.itemsPerPage;
        })
        .catch(error => {
            console.error('Error fetching NFT transfers:', error);
        })
        .finally(() => {
            loading.value = false;
        });
};
</script>
