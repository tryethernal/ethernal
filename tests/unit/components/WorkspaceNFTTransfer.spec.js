import WorkspaceNFTTransfer from '@/components/WorkspaceNFTTransfer.vue';

const stubs = ['TokenTransfers', 'VProgressLinear'];

describe('WorkspaceNFTTransfer.vue', () => {
    beforeEach(() => {
        vi.spyOn(console, 'error').mockImplementation(() => {});
    });

    afterEach(() => {
        vi.restoreAllMocks();
    });

    it('should render the component with loading state', async () => {
        vi.spyOn(server, 'getWorkspaceTokenTransfers').mockResolvedValueOnce({
            data: {
                items: [],
                count: 0
            }
        });

        const wrapper = mount(WorkspaceNFTTransfer, {
            global: {
                stubs
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render the component with data', async () => {
        const mockTransfers = [
            {
                transactionHash: '0x123',
                methodDetails: 'Transfer',
                timestamp: '2024-01-01',
                blockNumber: '100',
                src: '0xabc',
                dst: '0xdef',
                type: 'ERC721',
                token: 'NFT#1'
            }
        ];

        vi.spyOn(server, 'getWorkspaceTokenTransfers').mockResolvedValueOnce({
            data: {
                items: mockTransfers,
                count: 1
            }
        });

        const wrapper = mount(WorkspaceNFTTransfer, {
            global: {
                stubs
            }
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should handle sorting and pagination', async () => {
        const mockOptions = {
            page: 2,
            itemsPerPage: 10,
            sortBy: [{ key: 'timestamp', order: 'desc' }]
        };

        const mockTransfers = [
            {
                transactionHash: '0x123',
                methodDetails: 'Transfer',
                timestamp: '2024-01-01',
                blockNumber: '100',
                src: '0xabc',
                dst: '0xdef',
                type: 'ERC721',
                token: 'NFT#1'
            }
        ];

        const serverSpy = vi.spyOn(server, 'getWorkspaceTokenTransfers')
            .mockResolvedValueOnce({
                data: {
                    items: mockTransfers,
                    count: 1
                }
            });

        const wrapper = mount(WorkspaceNFTTransfer, {
            global: {
                stubs
            }
        });

        await flushPromises();
        await wrapper.findComponent({ name: 'TokenTransfers' }).vm.$emit('update:options', mockOptions);
        await flushPromises();

        expect(serverSpy).toHaveBeenCalledWith({
            page: mockOptions.page,
            limit: mockOptions.itemsPerPage,
            orderBy: mockOptions.sortBy[0].key,
            order: mockOptions.sortBy[0].order,
            tokenTypes: ['erc721', 'erc1155']
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should handle API errors', async () => {
        vi.spyOn(server, 'getWorkspaceTokenTransfers').mockRejectedValueOnce(new Error('API Error'));

        const wrapper = mount(WorkspaceNFTTransfer, {
            global: {
                stubs
            }
        });

        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot();
    });
}); 