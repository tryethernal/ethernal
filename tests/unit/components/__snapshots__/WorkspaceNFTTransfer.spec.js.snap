// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WorkspaceNFTTransfer.vue > should handle API errors 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <h2 class="text-h6 font-weight-medium">NFT Transfers</h2>
  <hr class="v-divider v-theme--light my-4" aria-orientation="horizontal" role="separator">
  <div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
    <!---->
    <div class="v-card__loader">
      <v-progress-linear-stub active="false" buffervalue="0" clickable="false" height="2" indeterminate="true" max="100" modelvalue="0" reverse="false" stream="false" striped="false" roundedbar="false" location="top" tile="false" tag="div"></v-progress-linear-stub>
    </div>
    <!---->
    <!---->
    <div class="v-card-text">
      <token-transfers-stub transfers="" headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" loading="false" sortby="" count="0" address="" density="default" withtransactiondata="false" withtokendata="false" nodatatext="No NFT transfers found." withcount="false"></token-transfers-stub>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;

exports[`WorkspaceNFTTransfer.vue > should handle sorting and pagination 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <h2 class="text-h6 font-weight-medium">NFT Transfers</h2>
  <hr class="v-divider v-theme--light my-4" aria-orientation="horizontal" role="separator">
  <div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
    <!---->
    <div class="v-card__loader">
      <v-progress-linear-stub active="false" buffervalue="0" clickable="false" height="2" indeterminate="true" max="100" modelvalue="0" reverse="false" stream="false" striped="false" roundedbar="false" location="top" tile="false" tag="div"></v-progress-linear-stub>
    </div>
    <!---->
    <!---->
    <div class="v-card-text">
      <token-transfers-stub transfers="[object Object]" headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" loading="false" sortby="" count="20" address="" density="default" withtransactiondata="false" withtokendata="false" nodatatext="No NFT transfers found." withcount="false"></token-transfers-stub>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;

exports[`WorkspaceNFTTransfer.vue > should render the component with data 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <h2 class="text-h6 font-weight-medium">NFT Transfers</h2>
  <hr class="v-divider v-theme--light my-4" aria-orientation="horizontal" role="separator">
  <div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
    <!---->
    <div class="v-card__loader">
      <v-progress-linear-stub active="false" buffervalue="0" clickable="false" height="2" indeterminate="true" max="100" modelvalue="0" reverse="false" stream="false" striped="false" roundedbar="false" location="top" tile="false" tag="div"></v-progress-linear-stub>
    </div>
    <!---->
    <!---->
    <div class="v-card-text">
      <token-transfers-stub transfers="" headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" loading="false" sortby="" count="0" address="" density="default" withtransactiondata="false" withtokendata="false" nodatatext="No NFT transfers found." withcount="false"></token-transfers-stub>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;

exports[`WorkspaceNFTTransfer.vue > should render the component with loading state 1`] = `
"<div class="v-container v-container--fluid v-locale--is-ltr">
  <h2 class="text-h6 font-weight-medium">NFT Transfers</h2>
  <hr class="v-divider v-theme--light my-4" aria-orientation="horizontal" role="separator">
  <div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
    <!---->
    <div class="v-card__loader">
      <v-progress-linear-stub active="false" buffervalue="0" clickable="false" height="2" indeterminate="true" max="100" modelvalue="0" reverse="false" stream="false" striped="false" roundedbar="false" location="top" tile="false" tag="div"></v-progress-linear-stub>
    </div>
    <!---->
    <!---->
    <div class="v-card-text">
      <token-transfers-stub transfers="" headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" loading="false" sortby="" count="0" address="" density="default" withtransactiondata="false" withtokendata="false" nodatatext="No NFT transfers found." withcount="false"></token-transfers-stub>
    </div>
    <!---->
    <!----><span class="v-card__underlay"></span>
  </div>
</div>"
`;
